feat(adr): addition of metrics collection ADR (#268)
edgexfoundry/edgex-docs: docs_src/design/adr/system-management/0006-Metrics-Collection.md

- Services will have configuration which allows EdgeX system managers to select which metrics are `on` or `off` by default - in other words providing the initial bootstrapping configuration that determines what metrics are collected and reported by default.
    - When a metric is turned `off` the service does not report the metric.  When a metric is turned `on` the service collects and sends the metric to the designated message topic.
    - Per REST API described above, the `on` and `off` control of the metrics collected can be changed during runtime of the service.
    - Metrics collection must be pushed to the designated message topic on some appointed schedule.  The schedule could be designated by configuration in the schedule service or done in a way similar to auto events in device services.  
        - *Do we want to dictate this or allow services to implement as they see fit?*

>         - *Also, should there me an interval per metric or a single interval for all metrics collected and pushed?*

Reviewer: typo is still there
should there **me** an interval

- Number of logins and login failures per service and within a given time
- Number of secrets accessed per service name
- Count of any accesses and failures to the data persistence layer
- Count of service start and restart attempts

### Design Proposal

- Each metric must have a unique name associate with it.  Because some metrics are reported from multiple services (such as service uptime), the name is not required to be unique across all services.  However, the name and the service name/key for a metric (see Bucket definition below) along with the created timestamp would uniquely identify a metric from a service at a designated time.
- All services will use/integrate go-mod-messaging (if not already integrated).
- A new metric DTO (Telemetry/Bucket) is required for services to structure the metric data.  The model/DTO will have similarities to the core event/reading model/DTO.
    - The DTO for the metric data will allow originating service information (the service key) to be provided in the DTO
    - The proposed message structure (DTO) for metrics (for inclusion in go-mod-core-contracts) is:
  
``` go

> type Telemtry struct {

Author: and in about 6 other spots.  good catch.

    Id          string            `json:"id,omitempty" codec:"id,omitempty"`  // UUID to identify the telemtry group
    Service     string            `json:"device,omitempty" codec:"device,omitempty"`  // originating service key
    Created     int64             `json:"created,omitempty" codec:"created,omitempty"`

>     Modified    int64             `json:"modified,omitempty" codec:"modified,omitempty"`

Author: I like that @cloudxxx8

    ValueType     string `json:"valueType,omitempty" codec:"valueType,omitempty"`  // metric value type
    Description   string `json:"name,omitempty" codec:"description,omitempty"`  // human readable details
}
```

!!! Note
    *For consideration, should we keep the metric data values simple (i.e. just a number field) since the metrics are just numbers and thereby avoid having to deal with types (pulling ValueType from above).  Even times could be sent as UTC number values.*

    *Also for consideration, can we simplify the structures?*

    - would the telemtry and bucket need an ID?  
    - would Created timestamp need to be put on the telemtry object?
    - would originating service need to be put in both telemetry and bucket objects?
    - would either Telemtry or Bucket be modified (necessitating a modified timestamp)?
    - Can the description be found in documentation and therefore left off of the Bucket?

- Proposed endpoint for the REST endpoint that responds with what metrics the service offers is:  /api/v2/metrics
- Body of the GET response would contain a JSON list of metrics (by metric name key) that are `on`

> - Body of the POST request would containe a JSON list of the metrics that are to be turned `on` (others are assumed to be turned `off`)

Author: Like the suggestion and changed accordingly.

    Name          string `json:"name,omitempty" codec:"name,omitempty"`  // metric name key
    Value         string `json:"value,omitempty" codec:"value,omitempty"` // metric value
    ValueType     string `json:"valueType,omitempty" codec:"valueType,omitempty"`  // metric value type
}
```

!!! Note
    Metrics are considered immutable and therefore not requiring a modified timestamp.

    *For consideration, should we keep the metric data values simple (i.e. just a number field) since the metrics are just numbers and thereby avoid having to deal with types (pulling ValueType from above).  Even times could be sent as UTC number values.*

    *Also for consideration, can we simplify the structures?*

    - would the telemetry and bucket need an ID?  
    - would origin timestamp need to be put on the telemetry object since it is on a Bucket?
    - would originating service need to be put in both telemetry and bucket objects?
    - Can the description be found in documentation and therefore left off of the Bucket?  *(@cloudxxx8 recommends removing.  Can add back in if there are dissenting opinions)*

> 

Reviewer 2: IMO this is re-inventing what is already available via packages like go-metrics. We should just use the raw metrics format from the chosen package.

Author: @lenny-intel - while I agree with your sentiment, in this case I don't know that it is practical for a couple of reasons.  The go-metrics package provides he means for collecting the telemetry data as key/values, but doesn't really provide the means to add the additional information (like service, tags, etc.) to the telemetry data when it is sent out/exported - which we would want to know who/where the telemetry came from (adding context)  Second of all, the way go-metrics works is that we are going to have to develop the means of getting metrics collected by go-metrics and then send or export them via MQTT.  So regardless of what information go-metrics has and what structure it uses (which isn't much), we are going to have to package that in an MQTT or other message bus message.  Since we are going to have to essentially pull the data, create a message, and push that message to a bus, it would seem to me that we want to create a message structure that adds the context others will need and make it look similar to our event/readings but for telemetry.

Author: per monthly architect's meeting - agreed to make the structure flat and allow user to format/structure the data as necessary. All that is needed is key/value pair with tags array.

The metric name must be unique for that service.  Because some metrics are reported from multiple services (such as service uptime), the name is not required to be unique across all services.

All information (keys, values, tags, etc.) is in string format and placed in a JSON array within the message body.  Here are some example representations:

**Example metric message body with a single value**
{"name":"service-up", "value":"120", "timestamp":"1602168089665570000", "tags":{"service":"coredata","uom":"days","type":"int64"}}

**Example metric message body with multiple values**
{"name":"api-requests", "value":"24", "timestamp":"1602168089665570001", "tags":{"service":"coredata","uom":"count","type":"int64", "mean":"0.0665", "rate1":"0.111", "rate5":"0.150","rate15":"0.111"}}

!!! Note
    The key or metric name must be unique when using go-metrics as it requires the metric name to be unique per the registry.  Metrics are considered immutable.

#### REST endpoints

!!! Note
    Again, these REST endpoints are per service and meant to control which metrics for a service are turned on or off.  The REST endpoints do not provide metrics data.

> - Proposed endpoint for the REST endpoint that responds with what metrics the service offers is:  /api/v2/metrics

Author: Good catch/point Lenny.  How about /api/v2/telemetry in order to avoid any confusion?

> - Body of the GET response would contain a JSON list of metrics (by metric name key) that are `on`

Author: Changed accordingly.  The GET response list will have to indicate which are on or off as well.

- Body of the PUT request would contain a JSON list of the metrics that are to be turned `on` (others are assumed to be turned `off`)
- Body of the PATCH request would contain a JSON list of the metrics that are to be turned `on` or `off` - leaving all other metrics unchanged

#### Configuration
- Configuration, not unlike that provided in core data or any device service, configuration will specify the message bus type and locations where the metrics messages should be sent.
- In fact, the message bus configuration will use (or reuse if the service is already using the message bus) the common message bus configuration as defined below.
- Metrics will be published to an /edgex/metrics/[service-name] topic where the service name will be added per service
- Common configuration for each service for message queue configuration - inclusive of metrics:

``` yaml
[MessageQueue]
Protocol = 'redis'  ## or 'tcp'
Host = 'localhost'
Port = 5573
Type = 'redis'  ## or 'mqtt'

> PublishTopicPrefix  = 'edgex/metrics' # /<service-name> will be added to this Publish Topic prefix

Reviewer 2: Device and Core data already have PublishTopicPrefix, so need an additional property for publishing metrics.
We could also add the metric name, allowing subscribers to filter for specific metrics if they wanted. This assume we publish the individual metrics separately, rather than all the service's metrics in one message.

```suggestion
MetricsPublishTopicPrefix  = 'edgex/metrics' # /<service-name>/<metric-name> will be added to this Publish Topic prefix

```

Author: Agreed.  Changed to /edgex/telemetry to keep it consistent with above and also added metric name.

Author: Also added TelemetryPublishTopicPrefix (vs PublishTopicPrefix which remains for device and core data).

  [MessageQueue.Optional]
  # Default MQTT Specific options that need to be here to enable environment variable overrides of them
  # Client Identifiers
  ClientId = "device-virtual"
  # Connection information
  Qos = "0" # Quality of Sevice values are 0 (At most once), 1 (At least once) or 2 (Exactly once)
  KeepAlive = "10" # Seconds (must be 2 or greater)
  Retained = "false"
  AutoReconnect = "true"
  ConnectTimeout = "5" # Seconds
  SkipCertVerify = "false" # Only used if Cert/Key file or Cert/Key PEMblock are specified
```

Additional configuration must be provided (in the service configuration.toml) to trigger the collection of telemetry from the metrics cache and sending it into the appointed message bus.
``` yaml

> [[Metrics.Collection]]

Reviewer 2: ```suggestion
[Metrics]

```

Author: Used Telemetry to align with other names and to avoid confusion with current system management

> Interval = "30s"

Author: moved

# EdgeX Metrics Collection

## Status

### Proposed

!!! Note
    This ADR initially incorporated metrics collection and control plane event processing.  The EdgeX architects felt the scope of the design was too large to cover under one ADR.  Control plane event processing will be covered under a separate ADR in the future.  For the purpose of distinction, metrics collection data and control plane events are defined below:

>     **Metric (or telemetry) data** is defined as the count or rate of some action, resource, or circumstance in the EdgeX instance or specific service.  Examples of metrics include:

Author: fixed

    - a service was stopped
    - service configuration has changed

    Note CPE should not be confused with core data Events.  Core data Events represent a collection (one or more) of sensor/device readings.  Core data Events represent sensing of some measured state of the physical world (temperature, vibration, etc.).  CPE represents the detection of some happening inside of the EdgeX software.

!!! Info
    Information in *italics* indicates topics for discussion without suggested proposal at this time.

## Context

[System Management services](../../../microservices/system-management/Ch_SystemManagement.md) (SMA and executors) currently provide a limited set of “metrics” to requesting clients (3rd party applications and systems external to EdgeX).  Namely, it provides requesting clients with service CPU and memory usage; both metrics about the resource utilization of the service itself versus metrics that are about what is happening inside of the service.  Arguably, the current system management metrics can be provided by the container engine and orchestration tools (example: by Docker engine) or by the underlying OS tooling.

!!! Info

>     The SMA has been deprecated and will be removed in a future, yet named, release.

Reviewer 3: This also doesn't render well. I'd suggest getting rid of the "!!! Info" and just using bold or italics to highlight.

```suggestion
    Note - The SMA has been deprecated and will be removed in a future, yet named, release.
```

Author: fixed

- sensor data collection (how much, how fast, etc.)
- command requests handled (how many, to which devices, etc.)
- sensor data transformation as it is done in application services (how fast, what is filtered, etc)
- sensor data export (how much is sent, how many exports have failed, etc. )
- API requests (how often, how quickly, how many success versus failed attempts, etc.)
- bootstrapping time (time to come up and be available to other services)
- activity processing time (amount of time it takes to perform a particular service function - such as respond to a command request)

### Definitions

**Metric (or telemetry) data** is defined as the count or rate of some action, resource, or circumstance in the EdgeX instance or specific service.  Examples of metrics include:

- the number of EdgeX Events sent from core data to an application service via message bus (or via device service to application service in Ireland and beyond)
- the number of requests on a service API
- the average time it takes to process a message through an application service
- The number of errors logged by a service

> The collection and dissemination of metric data will require internal service level instrumentation (relevant to that service) to capture and send data about relevant EdgeX operations.  EdgeX does not currently offer any service instrumentation.

Reviewer 3: Isn't there some measure of instrumentation today in order to support the albeit limited metrics available via SMA?

At this time, EdgeX will not persist the metric data (except as it may be retained as part of and message bus subsystem such as in an MQTT broker).  Consumers of metric data are responsible for persisting the data if needed, but this is external to EdgeX.  Persistence of metric information may be considered in the future based on requirements and adopter demand for such a feature.

In general, EdgeX metrics are meant to provide external applications and systems better information about what is happening "inside" EdgeX services and the associated devices with which it communicates.

### Requirements

- Services will push specified metrics collected for that service to a specified (by configuration) message endpoint (as supported by the EdgeX  message bus implementation; currently either Redis Pub/Sub or MQTT implementations are supported)

>     - Each service will have configuration that specifies a message endpoint for the service metrics.  The metrics message topic communications may be secured or unsecured (just as application services provide the means to export to secured or unsecured message pipes today).

Reviewer 3: I think this approach has some of the same issues as the recent North/South ADR. Namely, our default internal message bus implementation is secure by default. While it's possible today to access the message bus via an application service, it's not possible to (easily) grant permission to external clients. Doing this properly requires Vault support.

So does this mean the only way to access metrics from an external system would be via an application service which would act as a proxy/bridge to an external message bus, or do we add support to every service to allow an external message bus?

> - All services must document what metrics they offer.

Reviewer 3: Where? In edgex-docs? In the service's git repo?

Author: and the configuration will need to specify which metrics are offered and whether they are on or off - which I've added to the configuration part of the document (which was part of the removed REST API stuff.

- All EdgeX services must implement a common metrics interface/contract that defines an API set about service metrics.
    - The metrics REST endpoints implemented on each service are not there to provide the metrics data, but to know what metrics the service provides, to know the the current state for each metric (`on` or `off`), and to provide a means to turn `on` or `off` the metrics collection.  The actual metric data will be provided, when collected, in messages to a message bus.
    - The interface would include the definition of a GET REST endpoint that responds with the metrics that the service offers (and whether that metric is currently `on` or `off`). The proposed endpoint format:  **/api/v2/telemetry**.

>     - The interface would define PUT and PATCH REST endpoints that allows the user to toggle between `on` and `off` for any metric.  The proposed endpoint format:  **/api/v2/telemetry**.  In this REST request body would be the list of metric names to be turned `on` (or those turned `on` or `off` in the case of PATCH).  It is assumed that those metrics not listed are to be turned `off` in the PUT call where PATCH request is explicit about which metrics are turned `on` or `off` and leaves others unchanged.

Reviewer 4: I find it hard to understand the benefits of a networking API to modify telemetry collection over a networking API to collect the required metrics on demand.

Author: Discussed in the monthly architect's meeting of 1/24/22 and decision to not provide a REST API at this time.  Configuration will all be via configuration file and any changes would be via Consul.  If Consul is not in use, then the configuration could be static.

    - When a metric is turned `off` the service does not report the metric.  When a metric is turned `on` the service collects and sends the metric to the designated message topic.
    - Per REST API described above, the `on` and `off` control of the metrics collected can be changed during runtime of the service.
    - Metrics collection must be pushed to the designated message topic on some appointed schedule.  The schedule would be designated by configuration in the schedule service or done in a way similar to auto events in device services.  
        - *Do we want to dictate this or allow services to implement as they see fit?*
            - Recommendation by @cloudxxx8 - have internal scheduler but provide option to use external scheduler.   The service then has to provide the API for the external to call for collection.  Suggest using a design similar to auto discovery in device services.
        - *Also, should there be an interval per metric or a single interval for all metrics collected and pushed?*
            - Recommendation by @cloudxxx8 to support interval per metric but to have one for all as default

> 

Reviewer 3: Scheduling seems a bit hand-wavy here.

If each metric only supports a boolean status, it's not clear what the behavior is when a metric is set to `on` and no scheduling has been configured?

In order to use the support-scheduler, we'd need to implement a REST API trigger which has been mentioned, but isn't called out as a requirement in this version of the ADR.

If we were to support an internal auto-event like scheduler, then my suggestion would be to make this part of the per-metric configuration, and I'd like to see it actually described.

- Avg response time (in milliseconds or appropriate unit of measure) on APIs
- Avg and Max request size

#### Core/Supporting

- Latency (measure of time) an event takes to get through core data
- Latency (measure of time) a command request takes to get to a device service
- Indication of health – that events are being processed during a configurable period
- Number of events in persistence
- Number of readings in persistence
- Number of validation failures (validation of device identification)
- Number of notification transactions
- Number of notifications handled
- Number of failed notification transmissions
- Number of notifications in retry status

#### Application Services

- Processing time for a pipeline; latency (measure of time) an event takes to get through an  application service pipeline
- DB access times
- How often are we failing export to be sent to db to be retried at a later time
- What is the current store and forward queue size
- How much data (size in KBs or MBs) of packaged sensor data is being sent to an endpoint (or volume)
- Number of invalid messages that triggered pipeline
- Number of events processed

#### Device Services

> - Number of devices managed by this DS

Reviewer 3: This is already possible to get from Core Metadata, and it also doesn't align with the type of metrics described earlier in this ADR.

> - Device Requests (which may be more informative than reading counts and rates)

Reviewer 3: Suggestions:
  - number of events
  - number of readings

!!! Note
    There may be additional specific metrics for each device service.  For example, the ONVIF device service may report number of times camera tampering was detected. 

#### Security

> Security metrics may be more difficult to ascertain as they are cross service metrics.  They may have to be dealt with per service.  Also, true threat detection based on metrics may be a feature best provided by 3rd party based on particular threats and security profile needs.

Reviewer 3: These seem out of scope, as they would require instrumentation of services we don't own.

- Number of API requests denied due to wrong access token (Kong) per service and within a given time
- Number of secrets accessed per service name
- Count of any accesses and failures to the data persistence layer

> - Count of service start and restart attempts

Author: final metrics collected is determined by each work group providing the service.  This list was suggested by each work group, but put a note in that says this is just a list for consideration and left to implementation (and probably iterated over with new metrics being added with each release).

### Design Proposal

#### Collect and Push Architecture

> Metric data will be collected and cached by each service.  At designated times (kicked off by configurable schedule), the service will collect telemetry data from the cache and push it to a designated message bus topic.

Reviewer 3: Is there a reason we chose a push rather than pull model?

#### Metrics Messaging

> Cached metric data, at the designated time, will be marshaled into a message and pushed to the pre-configured message bus topic.

Reviewer 3: Was there any consideration to batching?

Each metric message consists of several key/value pairs:
- a **required** name (the name of the metric) such as service-uptime
- a **required** value which is the telemetry value collected such as 120 as the number of hours the service has been up.
- a **required** timestamp is the time (in Epoch timestamp/milliseconds format) at which the data was collected (similar in nature to the origin of sensed data). 

> - an optional collection (array) of tags.  The tags are sets of key/value pairs of strings that provide amplifying information about the telemetry.  Tags may include:

Reviewer 3: Why the added complexity? I'm not sure I see the value of tagging individual metrics?

    - originating service name
    - unit of measure associated with the telemetry value
    - value type of the value
    - additional values when the metric is more than just one value (example: when using a histogram, it would include min, max, mean and sum values)

The metric name must be unique for that service.  Because some metrics are reported from multiple services (such as service uptime), the name is not required to be unique across all services.

All information (keys, values, tags, etc.) is in string format and placed in a JSON array within the message body.  Here are some example representations:

**Example metric message body with a single value**
{"name":"service-up", "value":"120", "timestamp":"1602168089665570000", "tags":{"service":"coredata","uom":"days","type":"int64"}}

**Example metric message body with multiple values**
{"name":"api-requests", "value":"24", "timestamp":"1602168089665570001", "tags":{"service":"coredata","uom":"count","type":"int64", "mean":"0.0665", "rate1":"0.111", "rate5":"0.150","rate15":"0.111"}}

!!! Note

>     The key or metric name must be unique when using go-metrics as it requires the metric name to be unique per the registry.  Metrics are considered immutable.

Reviewer 3: unique across an EdgeX instance or individual EdgeX service?

Reviewer 2: @tonyespy , unique to the individual service.

#### REST endpoints

!!! Note
    Again, these REST endpoints are per service and meant to control which metrics for a service are turned on or off.  The REST endpoints do not provide metrics data.

- Proposed endpoint for the REST endpoint that responds with what metrics the service offers is:  /api/v2/telemetry
- Body of the GET response would contain a JSON list of metrics (by metric name key) and an indication of which are `on` and which are `off`
- Body of the PUT request would contain a JSON list of the metrics that are to be turned `on` (others are assumed to be turned `off`)

> - Body of the PATCH request would contain a JSON list of the metrics that are to be turned `on` or `off` - leaving all other metrics unchanged

Reviewer 3: There's no mention of scheduling?

Reviewer 2: @tonyespy , see `interval` below which is not part of this API.

Author: Given there is no API left in the service to have the schedule service to call, we have to do this internally now.  I've suggested that we do the scheduling like autoevent does it in device services.

- Configuration, not unlike that provided in core data or any device service, configuration will specify the message bus type and locations where the metrics messages should be sent.
- In fact, the message bus configuration will use (or reuse if the service is already using the message bus) the common message bus configuration as defined below.
- Common configuration for each service for message queue configuration - inclusive of metrics:

``` yaml
[MessageQueue]
Protocol = 'redis'  ## or 'tcp'
Host = 'localhost'
Port = 5573
Type = 'redis'  ## or 'mqtt'
PublishTopicPrefix = "edgex/events/core" # standard and existing core or device topic for publishing  
  [MessageQueue.Optional]
  # Default MQTT Specific options that need to be here to enable environment variable overrides of them
  # Client Identifiers
  ClientId = "device-virtual"
  # Connection information
  Qos = "0" # Quality of Service values are 0 (At most once), 1 (At least once) or 2 (Exactly once)
  KeepAlive = "10" # Seconds (must be 2 or greater)
  Retained = "false"
  AutoReconnect = "true"
  ConnectTimeout = "5" # Seconds
  SkipCertVerify = "false" # Only used if Cert/Key file or Cert/Key PEMblock are specified
```

> Additional configuration must be provided (in the service configuration.toml) to 

Reviewer 3: There's no mention of scheduling other then the presence of `Interval` below.

Reviewer 2: @tonyespy.  the `Interval` is the scheduling.

- Trigger the collection of telemetry from the metrics cache and sending it into the appointed message bus.
- Metrics will be published to an /edgex/telemetry/[service-name]/[metric-name] topic where the service name will be added per service and the metric name per metric (allowing subscribers to filter by service or metric name)

``` yaml
[[Telemetry]]

> Interval = "30s"

Reviewer 3: How does `Interval` work? Does this apply to all enabled metrics for the service (i.e. they're all triggered every 30s)?

Reviewer 2: Yes, they are all collected (if enabled) per interval.

```

#### Library Support
Each service will now need go-mod-messaging support.
Each service would determine when and what metrics to collect and push to the message bus, but will use a common library choose for each EdgeX language supported (Go or C currently)

There may be a desire to add common functionality in support of the REST handlers for getting the list of supported metrics and for enabling/disabling the metrics collection.  This can be explored at implementation time.

Use of [go-metrics](https://github.com/rcrowley/go-metrics) (a GoLang library to publish application metrics) would allow EdgeX to utilize (versus construct) a library utilized by over 7 thousand projects.  It provides the means to capture various types of metrics in a registry (a sophisticated map).  The metrics can then be published (`reported`) to a number of well known systems such as InfluxDB, Graphite, DataDog, and Syslog.  go-metrics is a Go library made from original Java package https://github.com/dropwizard/metrics.

A similar package would need to be selected for C.

** Considerations in the use of go-metrics **

> - This is a Golang only library.  Using this library would not provide with any package to use for the C services.  If there are expectations for parity between the services, this may be more difficult to achieve given the features of go-metrics.

Reviewer 3: I think it would be very wrong for us to not have parity between metrics produced by C services vs. Go services.

Reviewer 2: @tonyespy , as long as the C services produce the collect metrics in the same manner, we will have parity.

Author: parity but with different library us in C

- go-metrics will still require the EdgeX team to develop a bootstrapping apparatus to take the metrics configuration and register each of the metrics defined in the configuration in go-metrics.
- go-metrics would also require the EdgeX team to develop the means to periodically extract the metrics data from the registry and ship it via message bus (something the current go-metrics library does not do).

> - While go-metrics offers the ability for data to be reported to other subsystems, it would required EdgeX to expose these capabilities (possibly through APIs) if a user wanted to export to these subsystems in addition to the message bus.

Reviewer 3: I'm not sure what you mean by subsystems here? Can you clarify?

Reviewer 2: @tonyespy , should be other systems, such as Prometheus or InfluxDB

    - Operations:  update and get min, max, count, rate1, rate5, rate15, mean, percentile, sum and variance from the collection

``` Go
t := metrics.NewTimer()
t.Update(10)
time.Sleep(15 * time.Second)
t.Update(20)
time.Sleep(15 * time.Second)
t.Update(30)
time.Sleep(15 * time.Second)
t.Update(40)
time.Sleep(15 * time.Second)
fmt.Println((t.Max()))  // prints 40
fmt.Println(t.Min())    // prints 10
fmt.Println(t.Mean())   // prints 25
fmt.Println(t.Count())  // prints 4
fmt.Println(t.Sum())    // prints 100
fmt.Println(t.Percentile(0.25))  //prints 12.5
fmt.Println(t.Variance())  //prints 125
fmt.Println(t.Rate1())  // prints 0.1116017821771607
fmt.Println(t.Rate5())  // prints 0.1755821073441404
fmt.Println(t.Rate15()) // prints 0.1913711954736821
fmt.Println(t.RateMean()) //prints 0.06665773963998162

```

!!! Note
    The go-metrics package does offer some variants of these like the GaugeFloat64 to hold 64 bit floats.

## Consequences

- Should there be a *global* configuration option to turn all metrics off/on?
    - * EdgeX doesn't yet have global config so this will have to be by service. *

> - Given the potential that each service publishes metrics to the same message topic, 0MQ is not implementation option unless each service uses a different 0MQ pipe (0MQ topics do not allow multiple publishers).  

Reviewer 3: Since we've deprecated ZMQ, shouldn't this bit be removed?

Reviewer 2: @tonyespy,  ZMQ is still an option, just deprecated, so IMO this should stay so folks know they they can not use ZMQ with these Metrics.

**Additional metrics collection configuration** to be provided include: 

- Trigger the collection of telemetry from the metrics cache and sending it into the appointed message bus.
- Define which metrics are available and which are turned `off` and `on`.  All are false by default.  The list of metrics can and likely will be different per service.  The keys in this list are the metric name.  True and false are used for `on` and `off` values.
- Specify the metrics topic prefix where metrics data will be published to (ex:  providing the prefix /edgex/telemetry/topic name where the service and metric name `[service-name]/[metric-name]` will be appended per metric (allowing subscribers to filter by service or metric name)

These metrics configuration options will be defined in the `Writable` area of `configuration.toml` so as to allow for dynamic changes to the configuration (when using Consul).  Specifically, the `[Writable].[Writable.Telemetry]` area will dictate metrics collection configuration like this:

``` yaml

> [[Writable]]
>     [[Writable.Telemetry]]
>     Interval = "30s"
>     TelemetryPublishTopicPrefix  = "edgex/telemetry" # /<service-name>/<metric-name> will be added to this Publish Topic prefix
>     Metrics = "{\"service-up\":\"false\", \"api-requests\":\"false\"}"  #all metrics will be off (or false) by default configuration

Reviewer 2: This [[ is Array syntax for TOML. List of Metrics should be a `map` not embedded JSON. ```suggestion
[Writable]
   [Writable.Telemetry]
   Interval = "30s"
   PublishTopicPrefix  = "edgex/telemetry" # /<service-name>/<metric-name> will be added to this Publish Topic prefix
   [Writable.Telemetry.Metrics ]
   service-up = "false"
   api-requests = "false"  #all metrics will be off (or false) by default configuration
```

Author: changed accordingly

    - The configuration will be placed in the `Writable` area.  When a user wishes to change the configuration dynamically (such as turning on/off a metric), then Consul's UI can be used to change it.
- Services will have configuration which indicates what metrics are available from the service.
- Services will have configuration which allows EdgeX system managers to select which metrics are `on` or `off` - in other words providing configuration that determines what metrics are collected and reported by default.
    - When a metric is turned `off` (the default setting) the service does not report the metric.  When a metric is turned `on` the service collects and sends the metric to the designated message topic.
    - Metrics collection must be pushed to the designated message topic on some appointed schedule.  The schedule would be designated by configuration and done in a way similar to auto events in device services.  

>     - A default interval schedule is configured to determine the time when all metrics are collected, but an interval for any specific metric can be scheduled and override the default generic schedule. 

Author: changed and added a note that this may be relooked in the future.

These metrics configuration options will be defined in the `Writable` area of `configuration.toml` so as to allow for dynamic changes to the configuration (when using Consul).  Specifically, the `[Writable].[Writable.Telemetry]` area will dictate metrics collection configuration like this:

``` yaml
[[Writable]]
    [[Writable.Telemetry]]
    Interval = "30s"
    TelemetryPublishTopicPrefix  = "edgex/telemetry" # /<service-name>/<metric-name> will be added to this Publish Topic prefix
    Metrics = "{\"service-up\":\"false\", \"api-requests\":\"false\"}"  #all metrics will be off (or false) by default configuration
```

!!! Info
    It was discussed that in future EdgeX releases, services may want separate message bus connections.  For example one for sensor data and one for metrics telemetry data.  This would allow the QoS and other settings of the message bus connection to be different. This would allow sensor data collection, for example, to be messaged with a higher QoS than that of metrics.  For the initial release of this feature, the service will use the same connection (and therefore configuration) for metrics telemetry as well as sensor data.  

> 

Reviewer 2: We could also enhance go-mod-messaging to allow setting QOS on a per topic basis to avoid multiple connections.

Author: Added additional alternate suggestion to the Info comment.
